cocos/2d/utils/text-utils.ts

```diff
@@ -224,27 +410,25 @@ export function safeMeasureText (ctx: CanvasRenderingContext2D, string: string,
 // _safeSubstring(a, 1, 2) === _safeSubstring(a, 1, 3) === '😉'
 // _safeSubstring(a, 2, 3) === _safeSubstring(a, 2, 4) === '🚗'
 function _safeSubstring (targetString, startIndex, endIndex?): string {
-    let newStartIndex = startIndex;
-    let newEndIndex = endIndex;
-    const startChar = targetString[startIndex];
-    // lowSurrogateRex
-    if (startChar >= '\uDC00' && startChar <= '\uDFFF') {
-        newStartIndex--;
+    let newStartIndex = getSymbolStartIndex(targetString, startIndex);
+    if (newStartIndex < startIndex) {
+        newStartIndex = getSymbolEndIndex(targetString, startIndex) + 1;
     }
+    let newEndIndex = endIndex;
+
     if (endIndex !== undefined) {
-        if (endIndex - 1 !== startIndex) {
-            const endChar = targetString[endIndex - 1];
-            // highSurrogateRex
-            if (endChar >= '\uD800' && endChar <= '\uDBFF') {
-                newEndIndex--;
-            }
-        } else if (startChar >= '\uD800' && startChar <= '\uDBFF') {
-            // highSurrogateRex
-            newEndIndex++;
+        endIndex = Math.max(0, endIndex - 1);
+        newEndIndex = getSymbolEndIndex(targetString, endIndex);
+        const newStartEndIndex = getSymbolStartIndex(targetString, endIndex);
+        if (newStartEndIndex < newStartIndex || (newStartEndIndex == newStartIndex && startIndex > newStartIndex)) {
+            newEndIndex = newStartIndex;
```

Comment:
This judgment is entered when '_safeSubstring(txt, 4, 3)' is called, but the empty string is returned. I don't understand the purpose of this? It should return an error?

Reply:
startIndex > endIndex is not considered，you can help me to add this judgment

the multi-byte substring needs to consider the substrings concatenated same as the original string，The way I do this is I pick the first byte as whether in the substring，For example
_safeSubstring(txt, 0, 3) + _safeSubstring(txt, 4, 5) + _safeSubstring(txt, 6, 17)) 
=_safeSubstring(txt, 0, 17)
so if startindex is not the first byte，and endindex and startindex as the same char，should return empty string

Reply:
Okay, got it.
I've finished optimizing it and can help check it when you have time
https://github.com/cocos/cocos-engine/pull/16738